### PR TITLE
High pop reduced MC processing mode.

### PR DESCRIPTION
**Pull request by MrsTonedOne**
The logic behind this is that at higher populations, byond ends up needing to do more at the end of the tick to update clients, that the mc and traditional sleep timers end up fighting for a very small amount of time left.
Increasing the MC's sleep time means its wakes up sooner in the tick. So it has more time to do things, even if they don't happen as often, and leaving every other tick free allows for sleeping CHECK_TICK task to wake up without finding the MC left them very little time to do things.

Admins have been regularly manually doing this by varediting the processing variable to 2, that i figured we should automate it.

for /tg/, i plan on raising the player count a bit, but they make decent defaults for the avg server.

:cl:
tweak: The MC will now reduce tick rate during high populations to keep it from fighting with byond for processing time.
config: Added config options to control MC tick rate
admin: Admins can no longer manually control the mc's tick rate by editing the MC's processing value, instead you will have to edit the config datum's values for high/low pop tick rates.
/:cl:

### CODE REVIEW

File: code/controllers/configuration/entries/config.dm [hunk 1]
```diff
@@ -348,4 +348,27 @@ CONFIG_DEF(number/error_msg_delay)	// How long to wait between messaging admins
...
+CONFIG_DEF(number/mc_tick_rate/high_pop_mc_mode_amount)
+	value = 65
+
+CONFIG_DEF(number/mc_tick_rate/disable_high_pop_mc_mode_amount)
+	value = 60
+
+CONFIG_TWEAK(number/mc_tick_rate)
+	abstract_type = /datum/config_entry/number/mc_tick_rate
+
+CONFIG_TWEAK(number/mc_tick_rate/ValidateAndSet(str_val))
```
Comment (MrStonedOne) [inline code/controllers/configuration/entries/config.dm]: @Cyberboss I started pushing the boundaries of the config system.

**Comment by kevinz000** [comment]
don't lots of things depend on how fast this processes?

**Comment by MrStonedOne** [comment]
Yes, but once you get to high enough population, we can't afford to be processing that fast. The MC and byond and TICK_CHECK tasks start fighting for cpu time in a way that isn't productive for anybody.

You could fix this for flight suits by making them detect firing late (of any kind) and double move, kind of like you did with projectiles. as long as you don't do it for the first few moves after changing directions or starting to move, it shouldn't impact maneuverability too much

**Comment by kevinz000** [comment]
oof.

File: code/controllers/configuration/entries/config.dm [hunk 2]
```diff
@@ -348,4 +348,27 @@ CONFIG_DEF(number/error_msg_delay)	// How long to wait between messaging admins
...
+	. = ..()
+	if (. && Master)
+		Master.UpdateTickRate()
```
Comment (Cyberboss) [inline code/controllers/configuration/entries/config.dm]: Master is created before config so it's probably better to let this runtime if it doesn't exist

File: code/controllers/configuration/entries/config.dm (hunk 1 above)
Comment (Cyberboss) [reply code/controllers/configuration/entries/config.dm]: Debating whether or not this belongs here or in config_entry.dm....

Comment (MrStonedOne) [reply code/controllers/configuration/entries/config.dm]: Code organization is a balance between making a file so big it's hard to navigate and making code so split up is hard to find all the pieces.

If anything these should be split into category files of related things, sub-typed accordingly with overrides with them.

**Comment by Iamgoofball** [comment]
Maybe move all fastprocess processing to regular obj processing during highpop?

**Pull request merged**